File: prototype/node/src/app/core/escrow/payout-period.test.ts
import { test } from 'node:test'
import assert from 'node:assert/strict'
import { payoutPeriodEndingBefore, payoutPeriodEndsAt, payoutPeriodLabel } from './payout-period.ts'

test('a Monday settles the week that just ended', () => {
  const period = payoutPeriodEndingBefore(new Date('2026-08-24T09:00:00Z'))

  assert.deepEqual(period, { firstDay: '2026-08-17', lastDay: '2026-08-23' })
})

test('every day of a week settles the same period', () => {
  const period = payoutPeriodEndingBefore(new Date('2026-08-28T23:00:00Z'))

  assert.deepEqual(period, { firstDay: '2026-08-17', lastDay: '2026-08-23' })
})

test('a Sunday still belongs to the week it is closing', () => {
  const period = payoutPeriodEndingBefore(new Date('2026-08-30T09:00:00Z'))

  assert.deepEqual(period, { firstDay: '2026-08-17', lastDay: '2026-08-23' })
})

test('the period ends with the last instant of its last day', () => {
  const period = payoutPeriodEndingBefore(new Date('2026-08-24T09:00:00Z'))

  assert.equal(payoutPeriodEndsAt(period).toISOString(), '2026-08-23T23:59:59.999Z')
})

test('it labels itself with both ends', () => {
  const period = payoutPeriodEndingBefore(new Date('2026-08-24T09:00:00Z'))

  assert.equal(payoutPeriodLabel(period), '2026-08-17 to 2026-08-23')
})

File: prototype/rails/src/test/models/payout_period_test.rb
require "test_helper"

class PayoutPeriodTest < ActiveSupport::TestCase
  test "a monday settles the week that just ended" do
    period = PayoutPeriod.ending_before(Time.utc(2026, 8, 24, 9, 0, 0))

    assert_equal Date.new(2026, 8, 17), period.first_day
    assert_equal Date.new(2026, 8, 23), period.last_day
  end

  test "every day of a week settles the same period" do
    period = PayoutPeriod.ending_before(Time.utc(2026, 8, 28, 23, 0, 0))

    assert_equal Date.new(2026, 8, 17), period.first_day
    assert_equal Date.new(2026, 8, 23), period.last_day
  end

  test "a sunday still belongs to the week it is closing" do
    period = PayoutPeriod.ending_before(Time.utc(2026, 8, 30, 9, 0, 0))

    assert_equal Date.new(2026, 8, 17), period.first_day
    assert_equal Date.new(2026, 8, 23), period.last_day
  end

  test "it takes a date as readily as a time" do
    assert_equal Date.new(2026, 8, 17), PayoutPeriod.ending_before(Date.new(2026, 8, 24)).first_day
  end

  test "the period ends with the last second of its last day" do
    period = PayoutPeriod.ending_before(Time.utc(2026, 8, 24, 9, 0, 0))

    assert_equal Time.utc(2026, 8, 23, 23, 59, 59), period.ends_at
  end

  test "it labels itself with both ends" do
    assert_equal "2026-08-17 to 2026-08-23", PayoutPeriod.ending_before(Time.utc(2026, 8, 24)).label
  end
end
